File: src/handlers/auth/forgotPassword.ts
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import middy from '@middy/core';
import httpJsonBodyParser from '@middy/http-json-body-parser';
import httpErrorHandler from '@middy/http-error-handler';
import {
  CognitoIdentityProviderClient,
  ForgotPasswordCommand,
  UserNotFoundException,
} from '@aws-sdk/client-cognito-identity-provider';
import { z } from 'zod';
import { logger } from '../../shared/powertools';
import { config } from '../../shared/config';

const cognito = new CognitoIdentityProviderClient({});

const ForgotPasswordSchema = z.object({
  email: z.string().email(),
});

const lambdaHandler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
  const parseResult = ForgotPasswordSchema.safeParse(event.body);
  if (!parseResult.success) {
    return res(400, { message: parseResult.error.issues[0].message });
  }

  const { email } = parseResult.data;
  logger.info('Forgot password request', { email });

  try {
    await cognito.send(new ForgotPasswordCommand({
      ClientId: config.userPoolClientId,
      Username: email,
    }));

    return res(200, { message: 'Password reset code sent to your email.' });
  } catch (err) {
    if (err instanceof UserNotFoundException) {
      // Return 200 to avoid leaking which emails are registered
      return res(200, { message: 'Password reset code sent to your email.' });
    }
    logger.error('Forgot password failed', { email, error: String(err) });
    throw err;
  }
};

export const handler = middy(lambdaHandler)
  .use(httpJsonBodyParser())
  .use(httpErrorHandler());

function res(statusCode: number, body: object): APIGatewayProxyResult {
  return {
    statusCode,
    headers: { 'Access-Control-Allow-Origin': '*', 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  };
}

File: src/handlers/auth/forgot_password.py
import json
import re

import boto3
from botocore.exceptions import ClientError

from src.shared.config import config
from src.shared.powertools import logger, tracer

cognito = boto3.client("cognito-idp")

_EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")


def _res(status_code: int, body: dict) -> dict:
    return {
        "statusCode": status_code,
        "headers": {"Access-Control-Allow-Origin": "*", "Content-Type": "application/json"},
        "body": json.dumps(body),
    }


@logger.inject_lambda_context
@tracer.capture_lambda_handler
def handler(event, context):
    try:
        body = json.loads(event.get("body") or "{}")
    except (json.JSONDecodeError, TypeError):
        return _res(400, {"message": "Invalid JSON body"})

    email = str(body.get("email", "")).strip()

    if not email or not _EMAIL_RE.match(email):
        return _res(400, {"message": "Invalid email address"})

    logger.info("Forgot password request", email=email)

    try:
        cognito.forgot_password(
            ClientId=config["user_pool_client_id"],
            Username=email,
        )
        return _res(200, {"message": "Password reset code sent to your email."})
    except ClientError as e:
        error_code = e.response["Error"]["Code"]
        if error_code == "UserNotFoundException":
            return _res(200, {"message": "Password reset code sent to your email."})
        logger.error("Forgot password failed", email=email, error=str(e))
        raise
